middleware.js

```diff
@@ -0,0 +1,211 @@
+            for (var currentPhase = 0, pashesTotal = pashesKeys.length; currentPhase < pashesTotal; currentPhase++) {
+                var phaseName = Middleware.PHASES[pashesKeys[currentPhase]];
+                if(!phases[phaseName]){
+                    phases[phaseName] = {
+                        before : [],
+                        during : [],
+                        after : []
+                    };
+                }
+            }
+		},
+
+        /**
+         * Add a middleware to a specified phase
+         * @param {Middleware} middleware
+         * @param {String} phase - check Middleware.PHASES
+         * @param {String} subPhase - check Middleware.SUBPHASES
+         * @returns {boolean}
+         */
+		add : function (middleware, phase, subPhase){
+			if (phases[phase]){
+				if (phases[phase][subPhase]){
+					phases[phase][subPhase].push(middleware);
+				}
+				else{
+					phases[phase][DEFAULT_SUBPHASE].push(middleware);
+				}
+                return true;
+			}else {
+                return false;
+            }
+		},
+
+        /**
+         * Executes the middlewares sequentially for the given phase/subphase
+         * @param {String} phase - phase to be executed, check Middleware.PHASES
+         * @param {String} subPhase - subphase to be executed, check Middleware.SUBPHASES
+         * @param {Object} params - params to be pass to all the middlewares
+         * @param {Function} [onSuccess] - callback success function
+         * @param {Function} [onError] - callback error function
+         */
+		run : function(phase, subPhase, params, onSuccess, onError) {
+			if (phases[phase] && phases[phase][subPhase]){
+				this._runAll(phases[phase][subPhase], params).then(function (result) {
+					log('Middleware: ran successfully '+ phase + ':' + subPhase, result);
+					if (typeof onSuccess === 'function'){
+						onSuccess(result);
+					}
+				}).catch(function (error) {
+                    log('Middleware: ran with errors '+ phase + ':' + subPhase, error);
+					if (typeof onError === 'function'){
+						onError(error);
+					}
+				});
+			}
+		},
+
+        /***
+         * Executes the middlewares sequentially
+         * @param {Middleware[]} middlewares - array of middlewares
+         * @param {Object} params -  params to be passed to all the middlewares
+         * @returns {Promise}
+         * @private
+         */
+		_runAll : function(middlewares, params) {
+            /** iterates over the middleware promises. */
+			var _self =  this;
+            return middlewares.reduce( function(prevMiddlewarePromise, currentMiddleware) {
+                return prevMiddlewarePromise.then( function (value) {
+                    return new Promise(function (resolve, reject) {
+                        var middlewarePromise = _self._getMiddlewarePromise(currentMiddleware, params, value);
+                        return middlewarePromise.then(
+                            function onResolved() {
```

Comment:
have you tried adding only `resolve` itself?
I think it should have the same effect as `function(){ resolve.apply(this, arguments);}`